### PR TITLE
update shareboost error msg / update shareboost nonzero feat num boundary

### PR DESCRIPTION
**Pull request by OXPHOS**
Dynamic-cast tested.

### CODE REVIEW

File: src/shogun/multiclass/ShareBoost.cpp [hunk 1]
```diff
@@ -53,8 +53,9 @@ bool CShareBoost::train_machine(CFeatures* data)
 
 	if (m_nonzero_feas <= 0)
 		SG_ERROR("Set a valid (> 0) number of non-zero features to seek before training\n")
-	if (m_nonzero_feas >= dynamic_cast<CDenseFeatures<float64_t>*>(m_features)->get_num_features())
-		SG_ERROR("It doesn't make sense to use ShareBoost with num non-zero features >= num features in the data\n")
+	if (m_nonzero_feas > dynamic_cast<CDenseFeatures<float64_t>*>(m_features)->get_num_features())
+		SG_ERROR("Number of non-zero features (%d) cannot be larger than number of features (%d) in the data\n",
+				m_nonzero_feas, dynamic_cast<CDenseFeatures<float64_t>*>(m_features)->get_num_features())
 
 	m_fea = dynamic_cast<CDenseFeatures<float64_t> *>(m_features)->get_feature_matrix();
```
Comment (karlnapf) [inline src/shogun/multiclass/ShareBoost.cpp]: if you cast twice, why not store the result and only cast once?

**Comment by OXPHOS** [comment]
@karlnapf  updated

File: src/shogun/multiclass/ShareBoost.cpp [hunk 2]
```diff
@@ -53,8 +54,9 @@ bool CShareBoost::train_machine(CFeatures* data)
 
 	if (m_nonzero_feas <= 0)
 		SG_ERROR("Set a valid (> 0) number of non-zero features to seek before training\n")
-	if (m_nonzero_feas >= dynamic_cast<CDenseFeatures<float64_t>*>(m_features)->get_num_features())
-		SG_ERROR("It doesn't make sense to use ShareBoost with num non-zero features >= num features in the data\n")
+	if (m_nonzero_feas > num_features)
+		SG_ERROR("Number of non-zero features (%d) cannot be larger than number of features (%d) in the data\n",
+				m_nonzero_feas, num_features)
 
 	m_fea = dynamic_cast<CDenseFeatures<float64_t> *>(m_features)->get_feature_matrix();
```
Comment (karlnapf) [inline src/shogun/multiclass/ShareBoost.cpp]: there are still two dynamic casts in here.
`dynamic_cast<CDenseFeatures<float64_t> *>(m_features)` This is what I meant, not the num_features (though that can stay also)

**Comment by OXPHOS** [comment]
@karlnapf ready

File: src/shogun/multiclass/ShareBoost.cpp [hunk 3]
```diff
@@ -43,6 +43,7 @@ SGVector<int32_t> CShareBoost::get_activeset()
 
 bool CShareBoost::train_machine(CFeatures* data)
 {
+	int32_t num_features = dynamic_cast<CDenseFeatures<float64_t>*>(m_features)->get_num_features();
```
Comment (karlnapf) [inline src/shogun/multiclass/ShareBoost.cpp]: Lol, sorry we are miscommunicating.
So here you have a `dynamic_cast<CDenseFeatures<float64_t>*>(m_features)`

and a few lines below you have `m_fea = dynamic_cast<CDenseFeatures<float64_t> *>(m_features)`
I am talking about only casting once, not about the int

Comment (OXPHOS) [inline src/shogun/multiclass/ShareBoost.cpp]: Ohhhhhhh sry I missed that

File: src/shogun/multiclass/ShareBoost.cpp [hunk 4]
```diff
@@ -43,6 +43,7 @@ SGVector<int32_t> CShareBoost::get_activeset()
 
 bool CShareBoost::train_machine(CFeatures* data)
 {
+	CDenseFeatures<float64_t> *fea = dynamic_cast<CDenseFeatures<float64_t>*>(m_features);
```
Comment (karlnapf) [inline src/shogun/multiclass/ShareBoost.cpp]: @c4goldsw @vigsterkr 
Crhis, here is another one of those float64 casts, which shouldnt be there.
@OXPHOS just fixed a minor thing here, but we should systematically get rid of such casts everywhere. Can probably grep for them and have a list somewhere

**Comment by karlnapf** [comment]
Minor, all good here and travis was green before.

**Pull request merged**